### PR TITLE
Adding exceptino filter and unit tests

### PR DESCRIPTION
**Pull request by zain-h**

### CODE REVIEW

**Review by haziqAhmed92 — CHANGES_REQUESTED** [review]
attach screenshots of passing tests

**Comment by AbdulBasit0044** [comment]
Screenshot of the tests passing.

image

**Comment by AbdulBasit0044** [comment]
> attach screenshots of passing tests

Done

File: src/app.module.ts [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
 import { MongooseModule } from '@nestjs/mongoose';
 
 import { AppController } from './app.controller';
 import { AppService } from './app.service';
 import { TasksModule } from './tasks/tasks.module';
 
 @Module({
-  imports: [MongooseModule.forRoot('mongodb://localhost/demo'), TasksModule],
+  imports: [
+    MongooseModule.forRoot('mongodb://localhost/demo'),
```
Comment (mHamza301) [inline src/app.module.ts]: if we have already made the env file why not importing the variable by assigning it to process.env and then importing it here instead of manually declaring the same mongo connection string at 2 places.

File: src/tasks/tasks.controller.spec.ts [hunk 2]
```diff
@@ -1,18 +1,100 @@
...
+  });
+
+  it('should create a user!', async () => {
+    try {
+      var actual = await tasksController.createTask({
+        email: 'testing123344@gmail.com',
+        age: 22,
+      });
+    } catch (err) {
+      console.log(err);
```
Comment (mHamza301) [inline src/tasks/tasks.controller.spec.ts]: I guess by default jest suppresses the console logs can you please put a screenshot of this over here?

File: src/tasks/tasks.controller.spec.ts [hunk 3]
```diff
@@ -1,18 +1,100 @@
...
   });
 
   it('should be defined', () => {
-    expect(controller).toBeDefined();
+    expect(tasksController).toBeDefined();
+  });
+
+  it('should create a user!', async () => {
+    try {
+      var actual = await tasksController.createTask({
```
Comment (mHamza301) [inline src/tasks/tasks.controller.spec.ts]: Please don't use `vars` at all only use let and const. I hope you know about hoisting, this mistake shouldn't be repeated.

File: src/tasks/tasks.controller.spec.ts [hunk 4]
```diff
@@ -1,18 +1,100 @@
...
+      var actual = await tasksController.getTasks();
+    } catch (err) {
+      console.log(err);
+    }
+    expect(actual).toEqual([]);
+  });
+
+  it('should update the task!', async () => {
+    try {
+      const taskId = '1234';
```
Comment (mHamza301) [inline src/tasks/tasks.controller.spec.ts]: Can't we use uuid here instead of manually putting ids?

File: src/tasks/tasks.service.spec.ts [hunk 5]
```diff
@@ -15,4 +25,32 @@ describe('TasksService', () => {
   it('should be defined', () => {
     expect(service).toBeDefined();
```
Comment (mHamza301) [inline src/tasks/tasks.service.spec.ts]: Why we are always checking for the services to be defined?

File: test/task.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    app = moduleFixture.createNestApplication();
+    await app.init();
+  });
+
+  it('/ (GET)', () => {
+    return request(app.getHttpServer()).get('/tasks').expect(200);
+  });
+
+  it('/ (GETBYID)', () => {
```
Comment (mHamza301) [inline test/task.e2e-spec.ts]: Instead of getting by id over here, it should be better that we first run a post request and then get. This will free us from the burden of always storing a same id in db.

**Comment by mHamza301** [comment]
I have seen the test, but these only work for the positive conditions and not for the negative condition (edge cases) please update to ensure we have false cases as well

File: src/app.module.ts (hunk 1 above)
Comment (AbdulBasit0044) [reply src/app.module.ts]: Done now, i did tried to do that but I was not able to achieve it before, but it turns out that the order of importing modules matter in the import statement, but now it's resolved.

File: src/tasks/tasks.controller.spec.ts (hunk 3 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.controller.spec.ts]: Updated

File: src/tasks/tasks.controller.spec.ts (hunk 2 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.controller.spec.ts]: image

File: src/tasks/tasks.controller.spec.ts (hunk 4 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.controller.spec.ts]: updated with the uuid.

File: src/tasks/tasks.service.spec.ts (hunk 5 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.service.spec.ts]: this is the default test generated when we create a test file, so i thought to leave it as it is.

File: test/task.e2e-spec.ts (hunk 6 above)
Comment (AbdulBasit0044) [reply test/task.e2e-spec.ts]: Its a great idea, i did not think about it. Updated now accordingly.

**Comment by AbdulBasit0044** [comment]
> 

Added some edge cases for the false positive test cases.

File: src/tasks/tasks.service.spec.ts [hunk 7]
```diff
@@ -15,4 +25,32 @@ describe('TasksService', () => {
...
   });
+
+  it('should call createTask method', () => {
+    const dto = new CreateTaskDto();
+    const createTaskSpy = jest.spyOn(service, 'createTask');
+    service.createTask(dto);
+    expect(createTaskSpy).toHaveBeenCalledWith(dto);
+  });
+
+  it('should call getTasks method with expected param', async () => {
```
Comment (haziqAhmed92) [inline src/tasks/tasks.service.spec.ts]: remove async if you are not using await or promise.then() or catch()

File: src/tasks/tasks.service.spec.ts [hunk 8]
```diff
@@ -15,4 +25,32 @@ describe('TasksService', () => {
...
+    const getTasksSpy = jest.spyOn(service, 'getTasks');
+    service.getTasks();
+    expect(getTasksSpy).toHaveBeenCalledWith();
+  });
+
+  it('should call findOneTask method with expected param', async () => {
```
Comment (haziqAhmed92) [inline src/tasks/tasks.service.spec.ts]: remove async if you are not using await or promise.then() or catch()

File: src/tasks/tasks.service.spec.ts [hunk 9]
```diff
@@ -15,4 +25,32 @@ describe('TasksService', () => {
...
+    const findOneTaskSpy = jest.spyOn(service, 'getTaskById');
+    const taskId: string = '';
+    service.getTaskById(taskId);
+    expect(findOneTaskSpy).toHaveBeenCalledWith(taskId);
+  });
+
+  it('should call updateTask method with expected params', async () => {
```
Comment (haziqAhmed92) [inline src/tasks/tasks.service.spec.ts]: remove async if you are not using await or promise.then() or catch()

File: src/tasks/tasks.service.ts [hunk 10]
```diff
@@ -16,7 +17,8 @@ export class TasksService {
     return this.tasksRepository.find({});
   }
 
-  async createTask(email: string, age: number): Promise<Task> {
+  async createTask(createTaskDto: CreateTaskDto): Promise<Task> {
+    const {email, age} = createTaskDto;
     return this.tasksRepository.create({
```
Comment (haziqAhmed92) [inline src/tasks/tasks.service.ts:22]: same as above

File: test/task.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import * as request from 'supertest';
+import { TasksModule } from '../src/tasks/tasks.module';
+import { TasksService } from '../src/tasks/tasks.service';
+import { TasksRepository } from '../src/tasks/tasks.repository';
+import { AppModule } from '../src/app.module';
+
+describe('AppController (e2e)', () => {
+  let app: INestApplication;
+  const mockTasksRepository = {};
+
```
Comment (haziqAhmed92) [inline test/task.e2e-spec.ts:15]: `let request: () => supertest.SuperTest<supertest.Test>`
add this line

File: test/task.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      providers: [
+        TasksService,
+        {
+          provide: TasksRepository,
+          useValue: mockTasksRepository,
+        },
+      ],
+    }).compile();
+
+    app = moduleFixture.createNestApplication();
```
Comment (haziqAhmed92) [inline test/task.e2e-spec.ts:28]: ` request = () => supertest(app.getHttpServer())`
add this line 
now you dont have to call app.getHttpServer() every time

**Review by haziqAhmed92 — CHANGES_REQUESTED** [review]

File: src/tasks/tasks.service.ts (hunk 10 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.service.ts:22]: Done

File: src/tasks/tasks.service.spec.ts (hunk 9 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.service.spec.ts]: Done

File: src/tasks/tasks.service.spec.ts (hunk 8 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.service.spec.ts]: Done

File: src/tasks/tasks.service.spec.ts (hunk 7 above)
Comment (AbdulBasit0044) [reply src/tasks/tasks.service.spec.ts]: Done

File: src/tasks/tasks.controller.spec.ts (hunk 2 above)
Comment (mHamza301) [reply src/tasks/tasks.controller.spec.ts]: This is not what I asked for run the code and make sure it goes to the catch block I want to see the error up on console.

File: src/tasks/tasks.service.spec.ts (hunk 5 above)
Comment (mHamza301) [reply src/tasks/tasks.service.spec.ts]: @haziqAhmed92 what are your views on this?

Comment (haziqAhmed92) [reply src/tasks/tasks.service.spec.ts]: it will always be defined, cause you are importing it, otherwise the project would not run. 
please remove them as it is of no importance

Comment (AbdulBasit0044) [reply src/tasks/tasks.service.spec.ts]: Removed.
